### PR TITLE
Support for PowerShell CLM

### PR DESCRIPTION
**Pull request by DenisNikulin5**
This PR is a copy of https://github.com/microsoft/azure-pipelines-agent/pull/4502 with minor changes

Pull request for support of Constrained Language Mode for the ADO pipeline agent.
This refactors the code out of the CLI and into a script that can be signed, verified and worked with using WDAC and CLM.

Before we used a code of block inside PowerShell3Handler which was not allowed in constrained language mode, and it was a root cause of this error https://github.com/microsoft/azure-pipelines-agent/issues/4492

Testing: manually

### CODE REVIEW

File: src/Agent.Worker/Handlers/PowerShell3Handler.cs
Comment (KonstantinTyukalov) [inline src/Agent.Worker/Handlers/PowerShell3Handler.cs:1]: Please wrap this new logic by feature flag

File: src/Agent.Worker/Handlers/PowerShell3Handler.cs [hunk 1]
```diff
@@ -54,14 +54,26 @@ public async Task RunAsync()
             // Resolve the VSTS Task SDK module definition.
             string scriptDirectory = Path.GetDirectoryName(scriptFile);
             string moduleFile = Path.Combine(scriptDirectory, @"ps_modules", "VstsTaskSdk", "VstsTaskSdk.psd1");
+
             ArgUtil.File(moduleFile, nameof(moduleFile));
 
             // Craft the args to pass to PowerShell.exe.
+            /* 
+            Updates from Tim Brigham
+            This used to be a whole block of powershell passed on the CLI, but has been simplified into a wrapper "Use-VstsTaskScript.ps1"
```
Comment (KonstantinTyukalov) [inline src/Agent.Worker/Handlers/PowerShell3Handler.cs:63]: > Use-VstsTaskScript.ps1

Start-AzpTask.ps1

**Review by KonstantinTyukalov — CHANGES_REQUESTED** [review]

File: src/Misc/layoutbin/powershell/Start-AzpTask.ps1 [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+since it is parameterized it can be signed and trusted for WDAC and CLM.
+This is the original code that has been repalced - 
+@". ([scriptblock]::Create('if ([Console]::InputEncoding -is [Text.UTF8Encoding] -and [Console]::InputEncoding.GetPreamble().Length -ne 0) {{ [Console]::InputEncoding = New-Object Text.UTF8Encoding $false }} if (!$PSHOME) {{ $null = Get-Item -LiteralPath ''variable:PSHOME'' }} else {{ Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1'')) ; Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1'')) }}')) 2>&1 | ForEach-Object {{ Write-Verbose $_.Exception.Message -Verbose }} ; Import-Module -Name '{0}' -ArgumentList @{{ NonInteractive = $true }} -ErrorAction Stop ; $VerbosePreference = '{1}' ; $DebugPreference = '{1}' ; Invoke-VstsTaskScript -ScriptBlock ([scriptblock]::Create('. ''{2}'''))",
+#>
+
+param ( 
+    [Parameter(mandatory = $true)]$name, # Argument 0 
+    [Parameter(mandatory = $true)]$DebugOption, # Argument 1
+    [Parameter(mandatory = $true)]$scriptBlockString	# Argument 2
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Start-AzpTask.ps1:12]: Script args should be in upper case.
Also pls add expected type

File: src/Misc/layoutbin/powershell/Start-AzpTask.ps1 [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    catch [System.Management.Automation.PSNotSupportedException] {
+        $status = "ConstrainedLanguage"
+    }
+
+    Remove-Item $tempFile 
+    return $status 
+}
+
+# This is the notes we want to display if CLM is active on the device by default.
+$verboseNotes = @"
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Start-AzpTask.ps1:41]: I think the most part of these notes can be moved to a separate .md doc and only the url can be given here

File: src/Misc/layoutbin/powershell/Start-AzpTask.ps1 [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+@". ([scriptblock]::Create('if ([Console]::InputEncoding -is [Text.UTF8Encoding] -and [Console]::InputEncoding.GetPreamble().Length -ne 0) {{ [Console]::InputEncoding = New-Object Text.UTF8Encoding $false }} if (!$PSHOME) {{ $null = Get-Item -LiteralPath ''variable:PSHOME'' }} else {{ Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1'')) ; Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1'')) }}')) 2>&1 | ForEach-Object {{ Write-Verbose $_.Exception.Message -Verbose }} ; Import-Module -Name '{0}' -ArgumentList @{{ NonInteractive = $true }} -ErrorAction Stop ; $VerbosePreference = '{1}' ; $DebugPreference = '{1}' ; Invoke-VstsTaskScript -ScriptBlock ([scriptblock]::Create('. ''{2}'''))",
+#>
+
+param ( 
+    [Parameter(mandatory = $true)]$name, # Argument 0 
+    [Parameter(mandatory = $true)]$DebugOption, # Argument 1
+    [Parameter(mandatory = $true)]$scriptBlockString	# Argument 2 
+)
+
+function GetCLM {
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Start-AzpTask.ps1:15]: `Get-ClmStatus`?

File: src/Misc/layoutbin/powershell/Start-AzpTask.ps1 [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    # This is new functionality to detect if we are running in a constrained language mode.
+    # This is only used to display debug data if the device is in CLM mode by default.
+    # This test is modeled after the traditional __PSSCRIPTPOLICYTEST_LQU1DAME file
+    # This creates a temporary file with a changing signature 
+    # which will return the execution context language mode. 
+    $tempFileGuid = New-Guid | Select-Object -Expand Guid 
+    $tempFile = "$($env:temp)\$($tempFileGuid).ps1"
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Start-AzpTask.ps1:22]: Should we use `AGENT_TEMPDIRECTORY` instead?

File: src/Misc/layoutbin/powershell/Start-AzpTask.ps1 [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    [Console]::InputEncoding = New-Object Text.UTF8Encoding $false 
+} 
+
+if (!$PSHOME) { 
+    $null = Get-Item -LiteralPath 'variable:PSHOME' 
+}
+else { 
+    Import-Module -Name ([System.IO.Path]::Combine($PSHOME, 'Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1')) 
+    Import-Module -Name ([System.IO.Path]::Combine($PSHOME, 'Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1')) 
+}
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Start-AzpTask.ps1:119]: What does this  `if ... ` do?
Should we fail the task instead, or file a warning?

File: src/Agent.Worker/Handlers/PowerShell3Handler.cs
Comment (DenisNikulin5) [reply src/Agent.Worker/Handlers/PowerShell3Handler.cs:1]: Done

File: src/Agent.Sdk/Knob/AgentKnobs.cs [hunk 7]
```diff
@@ -724,5 +724,13 @@ public class AgentKnobs
             "Show warning message on the OS which is not supported by .NET 8",
             new PipelineFeatureSource("Net8UnsupportedOsWarning"),
             new BuiltInDefaultKnobSource("true"));
+
+        public static readonly Knob UsePSScriptWrapper = new Knob(
+            nameof(UsePSScriptWrapper),
+            "Use PowerShell script wrapper to handle PowerShell ConstrainedLanguage mode.",
+            new RuntimeKnobSource("USE_PS_SCRIPT_WRAPPER"),
+            new EnvironmentKnobSource("USE_PS_SCRIPT_WRAPPER"),
+            new PipelineFeatureSource("UsePSScriptWrapper"),
```
Comment (KonstantinTyukalov) [inline src/Agent.Sdk/Knob/AgentKnobs.cs:733]: Do we expecting the user will toggle this feature manually?

File: src/Misc/layoutbin/powershell/Start-AzpTask.ps1 [hunk 8]
```diff
@@ -0,0 +1,84 @@
+<# 
+A PowerShell script that is used to invoke a VSTS task script. This script is used by the VSTS task runner to invoke the task script.
+This script replaces some legacy stuff in PowerShell3Handler.cs and turns it into a dedicated signed script. 
+since it is parameterized it can be signed and trusted for WDAC and CLM.
+#>
+
+param ( 
+    [Parameter(mandatory = $true)]
+    [string]$Name,
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Start-AzpTask.ps1:9]: What does this param mean? Can we name this param more in a more self-descriptive way?

**Review by KonstantinTyukalov — APPROVED** [review]

File: src/Agent.Sdk/Knob/AgentKnobs.cs (hunk 7 above)
Comment (DenisNikulin5) [reply src/Agent.Sdk/Knob/AgentKnobs.cs]: Removed unnecessary ones

**Comment by DenisNikulin5** [comment]
Recreated this PR https://github.com/microsoft/azure-pipelines-agent/pull/4923 since PR checks don't work with repo forks

**Pull request closed without merging**